### PR TITLE
Adnuntius: Pass targeting to ad server

### PR DESCRIPTION
**Pull request by mikael-lundin**
We didn't pass the targeting keys that you could send via the params in prebid. 

image

Taken from here: https://docs.prebid.org/dev-docs/bidders/adnuntius.html

This PR aims to change that. :)

### CODE REVIEW

**Comment by mikael-lundin** [comment]
Is it possible to have someone check on this?

**Comment by scr-oath** [comment]
# PR #4545 Review Comments - Adnuntius Targeting Support

## General Review Comment (Post as main PR comment)

Thanks for adding targeting support to the Adnuntius adapter! The overall design with the nested Targeting struct is clean and follows Go best practices. Here are my key findings from the golang-pro and code-reviewer analysis:

### Critical/High Priority Issues

1. **Field Reordering** (`openrtb_ext/imp_adnuntius.go`:5)
   - The `Network` field was moved from position 2 to position 5
   - This can cause issues with JSON comparison and test diffs
   - **Recommendation**: Avoid reordering existing fields in production APIs

2. **Redundant Nil Checks** (`adapters/adnuntius/adnuntius_utils.go`:195-213)
   - The pattern checks if slices/strings are non-empty before assignment
   - This is unnecessary since JSON `omitempty` tags handle zero values
   - Consider simplifying to just check if `Targeting` is non-nil, then directly assign all fields

3. **Test Coverage** (`targeting-fields.json`)
   - Currently only one comprehensive test with all fields populated
   - Missing edge cases: partial targeting, empty targeting object, nil targeting pointer
   - **Recommendation**: Add supplemental test files for these scenarios

### Medium Priority

4. **Missing Documentation**
   - The new `Targeting` struct lacks godoc comments
   - Should document the purpose and expected format of each field

### Positive Aspects ✅

- Good API design with nested `Targeting` struct
- Consistent with existing adapter patterns
- Backward compatible with `omitempty` tags
- Comprehensive happy path test

**Overall**: Approve with minor changes recommended above.

File: openrtb_ext/imp_adnuntius.go [hunk 1]
```diff
@@ -2,8 +2,15 @@ package openrtb_ext
 
 type ImpExtAdnunitus struct {
 	Auid      string `json:"auId"`
-	Network   string `json:"network"`
 	NoCookies bool   `json:"noCookies"`
 	MaxDeals  int    `json:"maxDeals"`
```
Comment (scr-oath) [inline openrtb_ext/imp_adnuntius.go:6]: The `Network` field has been moved from its original position. This reordering can cause issues with JSON comparison in tests and makes git history harder to track.

**Recommendation**: Keep existing fields in their original order and only append new fields at the end. Consider reverting this change unless there's a specific reason for the reordering.

```go
// Suggested order:
type ImpExtAdnunitus struct {
    Auid      string `json:"auId"`
    Network   string `json:"network"`    // Keep original position
    NoCookies bool   `json:"noCookies"`
    MaxDeals  int    `json:"maxDeals"`
    BidType   string `json:"bidType,omitempty"`
    Targeting struct {
        // ... targeting fields
    } `json:"targeting,omitempty"`
}
```

File: openrtb_ext/imp_adnuntius.go [hunk 2]
```diff
@@ -2,8 +2,15 @@ package openrtb_ext
...
+	Network   string `json:"network"`
 	BidType   string `json:"bidType,omitempty"`
+	Targeting struct {
+		Category            []string            `json:"c,omitempty"`
+		Segments            []string            `json:"segments,omitempty"`
+		Keywords            []string            `json:"keywords,omitempty"`
+		KeyValues           map[string][]string `json:"kv,omitempty"`
+		AdUnitMatchingLabel []string            `json:"auml,omitempty"`
```
Comment (scr-oath) [inline openrtb_ext/imp_adnuntius.go:14]: The new `Targeting` struct looks good but is missing documentation. Consider adding godoc comments to explain the purpose of each field:

```go
// Targeting contains optional ad targeting parameters for the Adnuntius ad server.
// All fields are optional and correspond to Adnuntius API targeting capabilities.
// See https://docs.prebid.org/dev-docs/bidders/adnuntius.html for details.
Targeting struct {
    // Category specifies content categories for ad targeting (IAB categories)
    Category            []string            `json:"c,omitempty"`

    // Segments contains audience segment identifiers for targeting
    Segments            []string            `json:"segments,omitempty"`

    // Keywords contains targeting keywords
    Keywords            []string            `json:"keywords,omitempty"`

    // KeyValues contains custom key-value pairs for advanced targeting
    KeyValues           map[string][]string `json:"kv,omitempty"`

    // AdUnitMatchingLabel specifies labels for ad unit matching
    AdUnitMatchingLabel []string            `json:"auml,omitempty"`
} `json:"targeting,omitempty"`
```

File: adapters/adnuntius/adnuntius_utils.go [hunk 3]
```diff
@@ -191,6 +191,27 @@ func generateAdUnit(imp openrtb2.Imp, adnuntiusExt openrtb_ext.ImpExtAdnunitus,
...
+		adUnit.Keywords = adnuntiusExt.Targeting.Keywords
+	}
+
+	if adnuntiusExt.Targeting.KeyValues != nil {
+		adUnit.KeyValues = adnuntiusExt.Targeting.KeyValues
+	}
+
+	if adnuntiusExt.Targeting.AdUnitMatchingLabel != nil {
+		adUnit.AdUnitMatchingLabel = adnuntiusExt.Targeting.AdUnitMatchingLabel
+	}
```
Comment (scr-oath) [inline adapters/adnuntius/adnuntius_utils.go:213]: The nil checking pattern here is more verbose than necessary. Since the struct fields have `omitempty` JSON tags, zero values (empty slices, empty strings, nil maps) will automatically be omitted from the JSON output.

**Current approach**:
```go
if adnuntiusExt.Targeting.Category != nil {
    adUnit.Category = adnuntiusExt.Targeting.Category
}
if adnuntiusExt.Targeting.Segments != nil {
    adUnit.Segments = adnuntiusExt.Targeting.Segments
}
// ... etc
```

**Suggested simplification**:
```go
if adnuntiusExt.Targeting.Category != nil {
    adUnit.Category = adnuntiusExt.Targeting.Category
    adUnit.Segments = adnuntiusExt.Targeting.Segments
    adUnit.Keywords = adnuntiusExt.Targeting.Keywords
    adUnit.KeyValues = adnuntiusExt.Targeting.KeyValues
    adUnit.AdUnitMatchingLabel = adnuntiusExt.Targeting.AdUnitMatchingLabel
}
```

This is cleaner and relies on `omitempty` to handle zero values during marshaling.

File: adapters/adnuntius/adnuntiustest/supplemental/targeting-fields.json
Comment (scr-oath) [inline adapters/adnuntius/adnuntiustest/supplemental/targeting-fields.json:1]: Great comprehensive test! To ensure robustness, consider adding supplemental test files for these edge cases:

1. **`targeting-fields-partial.json`** - Only some targeting fields populated:
```json
{
  "targeting": {
    "c": ["automotive"],
    "keywords": ["car", "tech"]
    // segments, kv, auml omitted
  }
}
```

2. **`targeting-fields-empty.json`** - Empty targeting object (should be omitted from request):
```json
{
  "targeting": {}
}
```

3. **`targeting-fields-none.json`** - No targeting field at all (backward compatibility test):
```json
{
  "auId": "123"
  // no targeting field
}
```

These tests will ensure the adapter handles all scenarios correctly and maintains backward compatibility.

File: adapters/adnuntius/adnuntius_utils.go (hunk 3 above)
Comment (mikael-lundin) [reply adapters/adnuntius/adnuntius_utils.go:213]: but that looks like (in my eyes) that if I send category it will try to add all the values that may not be present in the request? Some requests maybe includes category others do not. At least that's the reason for the verbose-ness but my understanding of Go might be lacking here. :)

File: adapters/adnuntius/adnuntiustest/supplemental/targeting-fields.json
Comment (mikael-lundin) [reply adapters/adnuntius/adnuntiustest/supplemental/targeting-fields.json:1]: Will do!
